Interop/Tensorflow/src/main/java/org/tribuo/interop/tensorflow/TensorFlowTrainer.java

```diff
@@ -0,0 +1,597 @@
+import org.tensorflow.proto.framework.GraphDef;
+import org.tensorflow.types.TFloat32;
+import org.tensorflow.types.family.TNumber;
+import org.tribuo.Dataset;
+import org.tribuo.Example;
+import org.tribuo.ImmutableFeatureMap;
+import org.tribuo.ImmutableOutputInfo;
+import org.tribuo.Output;
+import org.tribuo.Trainer;
+import org.tribuo.provenance.ModelProvenance;
+import org.tribuo.provenance.SkeletalTrainerProvenance;
+import org.tribuo.provenance.TrainerProvenance;
+
+import java.io.BufferedInputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Instant;
+import java.time.OffsetDateTime;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Trainer for Tensorflow. Expects the underlying Tensorflow graph to have named placeholders for
+ * the inputs, ground truth outputs and a named output operation. The output operation should be
+ * before any softmax or sigmoid non-linearities to allow the use of more optimized loss functions.
+ * <p>
+ * This trainer only works with graphs setup for minibatches. To recover single example training just use a batch size of 1.
+ * <p>
+ * This trainer uses the serialisation functionality in {@link TensorFlowUtil}, or a TF checkpoint.
+ * <p>
+ * N.B. Tensorflow support is experimental and may change without a major version bump.
+ */
+public final class TensorFlowTrainer<T extends Output<T>> implements Trainer<T> {
+
+    private static final Logger logger = Logger.getLogger(TensorFlowTrainer.class.getName());
+
+    /**
+     * The model format to emit.
+     */
+    public enum TFModelFormat {
+        /**
+         * Saves the model inside the Tribuo object, emits a {@link TensorFlowNativeModel}.
+         */
+        TRIBUO_NATIVE,
+        /**
+         * Saves the model state inside a TensorFlow checkpoint, emits a {@link TensorFlowCheckpointModel}.
+         */
+        CHECKPOINT;
+    }
+
+    @Config(mandatory=true,description="Path to the protobuf containing the graph.")
+    private Path graphPath;
+
+    private GraphDef graphDef;
+
+    @Config(description="Test time batch size.")
+    private int testBatchSize = 16;
+
+    @Config(mandatory=true,description="Name of the output operation before the loss.")
+    private String outputName;
+
+    @Config(description="Name of the init operation.")
+    private String initName = Init.DEFAULT_NAME;
+
+    @Config(mandatory=true,description="Feature extractor.")
+    private FeatureConverter<T> featureConverter;
+
+    @Config(mandatory=true,description="Response extractor.")
+    private OutputConverter<T> outputConverter;
+
+    @Config(description="Minibatch size.")
+    private int minibatchSize = 1;
```

Comment:
`trainMinibatchSize` is more explicit and remove potential confusion about why there are two minibatch sizes.

Reply:
Sure.